CLUEstering/alpaka/CLUE/CLUEAlgoAlpaka.h

```diff
@@ -0,0 +1,212 @@
+    int pointsPerTile_;
+
+    /* domain_t<Ndim> m_domains; */
+
+    // Buffers
+    std::optional<cms::alpakatools::device_buffer<Device, TilesAlpaka<Ndim>>> d_tiles;
+    std::optional<cms::alpakatools::device_buffer<Device, cms::alpakatools::VecArray<int32_t, max_seeds>>>
+        d_seeds;
+    std::optional<
+        cms::alpakatools::device_buffer<Device, cms::alpakatools::VecArray<int32_t, max_followers>[]>>
+        d_followers;
+
+    // Private methods
+    void init_device(Queue queue_);
+    void setup(const Points<Ndim>& h_points, PointsAlpaka<Ndim>& d_points, Queue queue_);
+
+    // Construction of the tiles
+    void calculate_tile_size(TilesAlpaka<Ndim>& h_tiles, const Points<Ndim>& h_points);
+  };
+
+  // Private methods
+  template <typename TAcc, uint8_t Ndim>
+  void CLUEAlgoAlpaka<TAcc, Ndim>::calculate_tile_size(TilesAlpaka<Ndim>& h_tiles,
+                                                       const Points<Ndim>& h_points) {
+    for (int i{}; i != Ndim; ++i) {
+      float tileSize;
+      float dimMax{*std::max_element(h_points.m_coords[i].begin(), h_points.m_coords[i].end())};
+      float dimMin{*std::min_element(h_points.m_coords[i].begin(), h_points.m_coords[i].end())};
+
+      VecArray<float, 2> temp;
+      temp.push_back_unsafe(dimMin);
+      temp.push_back_unsafe(dimMax);
+      h_tiles.min_max[i] = temp;
+      tileSize = (dimMax - dimMin) / h_tiles.nPerDim();
+
+      h_tiles.tile_size[i] = tileSize;
+    }
+  }
+
+  template <typename TAcc, uint8_t Ndim>
+  void CLUEAlgoAlpaka<TAcc, Ndim>::init_device(Queue queue_) {
+    d_tiles = cms::alpakatools::make_device_buffer<TilesAlpaka<Ndim>>(queue_);
+    d_seeds = cms::alpakatools::make_device_buffer<cms::alpakatools::VecArray<int32_t, max_seeds>>(queue_);
+    d_followers =
+        cms::alpakatools::make_device_buffer<cms::alpakatools::VecArray<int32_t, max_followers>[]>(queue_,
+                                                                                                   reserve);
+
+    // Copy to the public pointers
+    m_seeds = (*d_seeds).data();
+    m_followers = (*d_followers).data();
+  }
+
+  template <typename TAcc, uint8_t Ndim>
+  void CLUEAlgoAlpaka<TAcc, Ndim>::setup(const Points<Ndim>& h_points,
+                                         PointsAlpaka<Ndim>& d_points,
+                                         Queue queue_) {
+    // Create temporary tiles object
+    TilesAlpaka<Ndim> temp;
+    calculate_tile_size(temp, h_points);
+    temp.resizeTiles();
+
+    alpaka::memcpy(queue_, *d_tiles, cms::alpakatools::make_host_view(temp));
+    m_tiles = (*d_tiles).data();
+    alpaka::memcpy(
+        queue_, d_points.coords, cms::alpakatools::make_host_view(h_points.m_coords.data(), h_points.n));
+    alpaka::memcpy(
+        queue_, d_points.weight, cms::alpakatools::make_host_view(h_points.m_weight.data(), h_points.n));
+    alpaka::memset(queue_, (*d_seeds), 0x00);
+
+    // Define the working division
+    const Idx block_size{1024};
```

Comment:
Should we allow the user to modify the work division?

Reply:
Good idea. I could add it as an attribute for the class and bind its setter

Reply:
Solved